src/opnsense/mvc/app/library/OPNsense/Backup/Mailer.php

```diff
@@ -0,0 +1,230 @@
+                $return = self::sendEmail($config->system->remotebackup, $confdata);
+            }
+        }
+
+        return array($return);
+    }
+
+    /**
+     * Is this provider enabled
+     * @return boolean enabled status
+     */
+    public function isEnabled()
+    {
+        $cnf = Config::getInstance();
+        if ($cnf->isValid()) {
+            $config =$cnf->object();
+            return isset($config->system->remotebackup) && isset($config->system->remotebackup->MailEnabled)
+                && !empty($config->system->remotebackup->MailEnabled);
+        }
+        return false;
+    }
+
+    public function sendEmail($config, $confdata)
+    {
+        if (!`which gpg2`) {
+            $link = 'http://pkg.freebsd.org/freebsd:11:x86:64/latest/All/';
+
+            $dependencies = array(
+                'libgpg-error-1.32.txz', 'libgcrypt-1.8.3.txz', 'libksba-1.3.5.txz',
+                'libtasn1-4.13.txz', 'p11-kit-0.23.12.txz', 'libunistring-0.9.10.txz',
+                'libidn2-2.0.5.txz', 'libgpg-error-1.32.txz', 'libassuan-2.5.1.txz',
+                'tpm-emulator-0.7.4_2.txz', 'trousers-0.3.14_2.txz', 'pinentry-tty-1.1.0.txz',
+                'pinentry-1.1.0_1.txz', 'npth-1.6.txz', 'gnutls-3.5.18.txz', 'gnupg-2.2.9_1.txz'
+            );
+
+            foreach ($dependencies as $dependency) {
+                exec('pkg add ' . $link . $dependency . ' > /dev/null 2>&1');
+            }
+        }
+
+        $smtpUsername = $config->SmtpUsername;
+        $smtpPassword = $config->SmtpPassword;
+        $gpgPublicKey = $config->GpgPublicKey;
+        $gpgEmail     = $config->GpgEmail;
+
+        $date     = date('Y-m-d');
+        $hostname = gethostname();
+
+        $mail = new PHPMailer(true);
+        $mail->IsHTML(true);
+        $mail->IsSMTP();
+        $mail->SetFrom($gpgEmail);
+        $mail->AddAddress($config->Receiver);
+        $mail->Host    = $config->SmtpHost;
+        $mail->Port    = $config->SmtpPort;
+        $mail->Subject = $hostname . ' OPNsense config backup ' . $date;
+        $mail->Body    = $hostname . ' config backup file';
+
+        if ($config->SmtpSSL == "on") {
+            $mail->SMTPSecure = 'ssl';
+        }
+        $mail->SMTPOptions = array(
+            'ssl' => array(
+                'verify_peer' => false,
+                'verify_peer_name' => false,
+                'allow_self_signed' => true,
+            ),
+        );
+
+        if ($smtpUsername != "" && $smtpPassword != "") {
+            $mail->SMTPAuth = true;
+            $mail->Username = $smtpUsername;
+            $mail->Password = $smtpPassword;
+        }
+
+        $gpgPublicKeyFile = "key.asc";
+        if ($gpgPublicKey != "") {
+            file_put_contents($gpgPublicKeyFile, $gpgPublicKey);
+        }
+
+        exec('gpg2 --import < ' . $gpgPublicKeyFile);
+        exec('echo "' . $confdata . '" | gpg2 --trust-model always --batch --yes --output backup --encrypt --recipient ' . $gpgEmail);
```

Comment:
looks like high danger zone - the problem here is that it looks like it is vulnerable to command injection.

Reply:
should I try to find another way to do this or input validation would be enough?

Reply:
https://secure.php.net/manual/de/function.proc-open.php for the command calls and shellescapearg for the aguments.